File: src/typechecker/calls.rs
use super::*;

/// Pre-registers every top-level function's parameter and return types
/// into `self.functions`, without checking any body. Run before any
/// bodies are checked so that a call to a function defined later in the
/// file (or a recursive call to the function currently being checked)
/// already has a known signature to check argument types/arity against.
pub(super) fn collect_function_signatures(
    checker: &mut TypeChecker,
    program: &[Stmt],
) -> Result<(), TypeError> {
    for stmt in program {
        if let Stmt::FunctionDef {
            name,
            params,
            return_type,
            line,
            column,
            ..
        } = stmt
        {
            let mut param_types = Vec::new();
            for p in params {
                param_types.push(checker.resolve_type(&p.type_ann.name, p.line, p.column)?);
            }
            let return_type = match return_type {
                Some(t) => Some(checker.resolve_type(&t.name, *line, *column)?),
                None => None,
            };
            checker.functions.insert(
                name.clone(),
                FunctionSig {
                    param_types,
                    return_type,
                },
            );
        }
    }
    Ok(())
}

/// Type-checks array and pointer builtins if `callee` names one of them, returning
/// `Ok(None)` for any other callee so the caller falls through to user-defined
/// function lookup.
/// Type-checks a free function call `callee(args)`, resolving in order: the
/// `print` builtin (any args, yields `Nil`), the array/pointer builtins (via
/// `check_array_builtin`), then a user-defined function in `self.functions`
/// — checking argument count and each argument's type against the signature.
/// Yields the function's declared return type, or `Nil` if it declares none.
/// Split out of `check_expr`'s `Call` arm to keep that dispatch readable.
pub(super) fn check_call(
    checker: &mut TypeChecker,
    callee: &str,
    args: &[Expr],
    line: usize,
    column: usize,
) -> Result<Type, TypeError> {
    if callee == "print" {
        for a in args {
            super::exprs::check_expr(checker, a)?;
        }
        return Ok(Type::Nil);
    }
    if let Some(ty) = check_array_builtin(checker, callee, args, line, column)? {
        return Ok(ty);
    }
    let sig = checker
        .functions
        .get(callee)
        .cloned()
        .ok_or_else(|| TypeError {
            message: format!("undefined function `{callee}`"),
            line,
            column,
        })?;
    if args.len() != sig.param_types.len() {
        return Err(TypeError {
            message: format!(
                "function `{callee}` expects {} argument(s), found {}",
                sig.param_types.len(),
                args.len()
            ),
            line,
            column,
        });
    }
    for (arg, expected) in args.iter().zip(sig.param_types.iter()) {
        let arg_ty = super::exprs::check_expr(checker, arg)?;
        if arg_ty != *expected {
            return Err(TypeError {
                message: format!("argument to `{callee}` expects `{expected}`, found `{arg_ty}`"),
                line: arg.line(),
                column: arg.column(),
            });
        }
    }
    Ok(sig.return_type.unwrap_or(Type::Nil))
}

/// Shared arity/type-checking for any call-like site that already has a
/// resolved `FunctionSig` — user-defined top-level function calls,
/// instance method calls, and `.new` construction all funnel through
/// here. Checks `args.len()` matches `sig.param_types.len()` first (using
/// `what` — e.g. `"add"`, `"Hello#greet"`, `"Hello.new"` — to name the
/// callee in the error), then checks each argument expression and
/// requires its type to exactly equal the corresponding declared
/// parameter type (no implicit coercion).
pub(super) fn check_call_args(
    checker: &mut TypeChecker,
    what: &str,
    sig: &FunctionSig,
    args: &[Expr],
    line: usize,
    column: usize,
) -> Result<(), TypeError> {
    if args.len() != sig.param_types.len() {
        return Err(TypeError {
            message: format!(
                "`{what}` expects {} argument(s), found {}",
                sig.param_types.len(),
                args.len()
            ),
            line,
            column,
        });
    }
    for (arg, expected) in args.iter().zip(sig.param_types.iter()) {
        let arg_ty = super::exprs::check_expr(checker, arg)?;
        if arg_ty != *expected {
            return Err(TypeError {
                message: format!("argument to `{what}` expects `{expected}`, found `{arg_ty}`"),
                line: arg.line(),
                column: arg.column(),
            });
        }
    }
    Ok(())
}

/// Type-checks array and pointer builtins. Arity is checked once here from the
/// registry; the per-builtin arms below can then trust args have exactly the
/// right number of elements. Handles `len`/`push`/`get`/`set`/`pop` (array
/// builtins), `alloc`/`deref`/`set_deref`/`free` (pointer builtins), and `collect` (GC).
fn check_array_builtin(
    checker: &mut TypeChecker,
    callee: &str,
    args: &[Expr],
    line: usize,
    column: usize,
) -> Result<Option<Type>, TypeError> {
    let Some(builtin) = crate::builtins::lookup(callee) else {
        return Ok(None);
    };
    // Arity is checked once here from the registry; the per-builtin arms
    // below can then trust `args` has exactly `builtin.arity` elements.
    if args.len() != builtin.arity {
        return Err(TypeError {
            message: format!(
                "`{callee}` expects {} argument(s), found {}",
                builtin.arity,
                args.len()
            ),
            line,
            column,
        });
    }
    match callee {
        "len" => match super::exprs::check_expr(checker, &args[0])? {
            Type::Array(_) => Ok(Some(Type::Integer)),
            other => Err(TypeError {
                message: format!("`len` expects an array, found `{other}`"),
                line,
                column,
            }),
        },
        "push" => {
            let array_ty = super::exprs::check_expr(checker, &args[0])?;
            let value_ty = super::exprs::check_expr(checker, &args[1])?;
            match array_ty {
                Type::Array(elem) if *elem == value_ty => Ok(Some(Type::Nil)),
                Type::Array(elem) => Err(TypeError {
                    message: format!(
                        "`push` onto `Array<{elem}>` expects `{elem}`, found `{value_ty}`"
                    ),
                    line,
                    column,
                }),
                other => Err(TypeError {
                    message: format!("`push` expects an array, found `{other}`"),
                    line,
                    column,
                }),
            }
        }
        "get" => {
            let array_ty = super::exprs::check_expr(checker, &args[0])?;
            let index_ty = super::exprs::check_expr(checker, &args[1])?;
            if index_ty != Type::Integer {
                return Err(TypeError {
                    message: format!("`get` index must be Integer, found `{index_ty}`"),
                    line,
                    column,
                });
            }
            match array_ty {
                Type::Array(elem) => Ok(Some(*elem)),
                other => Err(TypeError {
                    message: format!("`get` expects an array, found `{other}`"),
                    line,
                    column,
                }),
            }
        }
        "set" => {
            let array_ty = super::exprs::check_expr(checker, &args[0])?;
            let index_ty = super::exprs::check_expr(checker, &args[1])?;
            let value_ty = super::exprs::check_expr(checker, &args[2])?;
            if index_ty != Type::Integer {
                return Err(TypeError {
                    message: format!("`set` index must be Integer, found `{index_ty}`"),
                    line,
                    column,
                });
            }
            match array_ty {
                Type::Array(elem) if *elem == value_ty => Ok(Some(Type::Nil)),
                Type::Array(elem) => Err(TypeError {
                    message: format!(
                        "`set` onto `Array<{elem}>` expects `{elem}`, found `{value_ty}`"
                    ),
                    line,
                    column,
                }),
                other => Err(TypeError {
                    message: format!("`set` expects an array, found `{other}`"),
                    line,
                    column,
                }),
            }
        }
        "pop" => match super::exprs::check_expr(checker, &args[0])? {
            Type::Array(elem) => Ok(Some(*elem)),
            other => Err(TypeError {
                message: format!("`pop` expects an array, found `{other}`"),
                line,
                column,
            }),
        },
        "alloc" => {
            let value_ty = super::exprs::check_expr(checker, &args[0])?;
            Ok(Some(Type::Pointer(Box::new(value_ty))))
        }
        "deref" => {
            let ptr_ty = super::exprs::check_expr(checker, &args[0])?;
            match ptr_ty {
                Type::Pointer(elem) => Ok(Some(*elem)),
                other => Err(TypeError {
                    message: format!("`deref` expects a pointer, found `{other}`"),
                    line,
                    column,
                }),
            }
        }
        "set_deref" => {
            let ptr_ty = super::exprs::check_expr(checker, &args[0])?;
            let value_ty = super::exprs::check_expr(checker, &args[1])?;
            match ptr_ty {
                Type::Pointer(elem) if *elem == value_ty => Ok(Some(Type::Nil)),
                Type::Pointer(elem) => Err(TypeError {
                    message: format!(
                        "`set_deref` into `Ptr<{elem}>` expects `{elem}`, found `{value_ty}`"
                    ),
                    line,
                    column,
                }),
                other => Err(TypeError {
                    message: format!("`set_deref` expects a pointer, found `{other}`"),
                    line,
                    column,
                }),
            }
        }
        "free" => {
            let ptr_ty = super::exprs::check_expr(checker, &args[0])?;
            match ptr_ty {
                Type::Pointer(_) => Ok(Some(Type::Nil)),
                other => Err(TypeError {
                    message: format!("`free` expects a pointer, found `{other}`"),
                    line,
                    column,
                }),
            }
        }
        // Mark-and-sweep GC over the pointer heap; yields the freed-slot count.
        "collect" => Ok(Some(Type::Integer)),
        _ => unreachable!("builtin `{callee}` is in the registry but has no typecheck arm"),
    }
}

File: ruby/lib/yara/typechecker/calls.rb
module Yara
  class TypeChecker
    private

    # Registers every top-level function's signature before any body is
    # checked, so calls may precede definitions or recurse. A return type
    # error points at the `def`.
    #
    # @param program [Array<AST::Node>]
    # @return [void]
    def collect_function_signatures(program)
      program.grep(AST::FunctionDef).each do |function|
        @functions[function.name] = signature_of(function)
      end
    end

    # @param function [AST::FunctionDef]
    # @return [Signature]
    def signature_of(function)
      params = function.params.map { |param| resolve_type(param.type_ann.name, param.line, param.column) }
      returns = function.return_type && resolve_type(function.return_type.name, function.line, function.column)
      Signature.new(params, returns)
    end

    # A free call resolves to `print` (any arguments), a registry builtin, or
    # a user function, in that order.
    #
    # @param callee [String]
    # @param args [Array<AST::Node>]
    # @param line [Integer]
    # @param column [Integer]
    # @return [Type]
    def check_call(callee, args, line, column)
      canonical = @vocabulary.canonical_builtin(callee)
      if canonical == "print"
        args.each { |arg| check_expr(arg) }
        return Type::NIL
      end
      return check_builtin(canonical, callee, args, line, column) if Builtins::ARITIES.key?(canonical)

      signature = @functions[callee] || type_error("type/undefined-function", [callee], line, column)
      check_arguments(callee, signature, args, line, column, "type/function-arity-mismatch")
      signature.return_type || Type::NIL
    end

    # Arity and exact argument types, shared by function, method and `.new`
    # calls; `what` names the callee in errors (`add`, `Hello#greet`,
    # `Hello.new`).
    #
    # @param what [String]
    # @param signature [Signature]
    # @param args [Array<AST::Node>]
    # @param line [Integer]
    # @param column [Integer]
    # @param arity_key [String] message catalog key for an arity mismatch
    # @return [void]
    def check_arguments(what, signature, args, line, column, arity_key = "type/call-arity-mismatch")
      if args.size != signature.param_types.size
        type_error(arity_key, [what, signature.param_types.size.to_s, args.size.to_s], line, column)
      end
      args.zip(signature.param_types).each do |arg, expected|
        actual = check_expr(arg)
        next if expected.accepts?(actual)

        type_error("type/argument-type-mismatch", [what, name_of(expected), name_of(actual)], arg.line, arg.column)
      end
    end

    # @param canonical [String] the builtin's English name
    # @param callee [String] the name as written, for errors
    # @param args [Array<AST::Node>]
    # @param line [Integer]
    # @param column [Integer]
    # @return [Type]
    def check_builtin(canonical, callee, args, line, column)
      arity = Builtins::ARITIES[canonical]
      if args.size != arity
        type_error("type/builtin-arity-mismatch", [callee, arity.to_s, args.size.to_s], line, column)
      end
      send("check_builtin_#{canonical}", args, line, column)
    end

    # @return [Type]
    def check_builtin_len(args, _line, _column)
      array = check_expr(args[0])
      return Type::INTEGER if array.kind == "Array"

      type_error("type/len-expects-array", [name_of(array)], args[0].line, args[0].column)
    end

    # Unlike `Array#push`, the value must match the element type exactly.
    #
    # @return [Type]
    def check_builtin_push(args, line, column)
      array = check_expr(args[0])
      value = check_expr(args[1])
      type_error("type/push-expects-array", [name_of(array)], line, column) if array.kind != "Array"
      return Type::NIL if array.inner == value

      type_error("type/push-onto-mismatch", [name_of(array.inner), name_of(value)], line, column)
    end

    # @return [Type]
    def check_builtin_get(args, line, column)
      array = check_expr(args[0])
      index = check_expr(args[1])
      type_error("type/get-index-not-integer", [name_of(index)], line, column) if index != Type::INTEGER
      return array.inner if array.kind == "Array"

      type_error("type/get-expects-array", [name_of(array)], line, column)
    end

    # @return [Type]
    def check_builtin_set(args, line, column)
      array = check_expr(args[0])
      index = check_expr(args[1])
      value = check_expr(args[2])
      type_error("type/set-index-not-integer", [name_of(index)], line, column) if index != Type::INTEGER
      type_error("type/set-expects-array", [name_of(array)], line, column) if array.kind != "Array"
      return Type::NIL if array.inner == value

      type_error("type/set-onto-mismatch", [name_of(array.inner), name_of(value)], line, column)
    end

    # @return [Type]
    def check_builtin_pop(args, _line, _column)
      array = check_expr(args[0])
      return array.inner if array.kind == "Array"

      type_error("type/pop-expects-array", [name_of(array)], args[0].line, args[0].column)
    end

    # @return [Type]
    def check_builtin_alloc(args, _line, _column)
      Type.pointer(check_expr(args[0]))
    end

    # @return [Type]
    def check_builtin_deref(args, line, column)
      pointer = check_expr(args[0])
      return pointer.inner if pointer.kind == "Pointer"

      type_error("type/deref-expects-pointer", [name_of(pointer)], line, column)
    end

    # @return [Type]
    def check_builtin_set_deref(args, line, column)
      pointer = check_expr(args[0])
      value = check_expr(args[1])
      type_error("type/set-deref-expects-pointer", [name_of(pointer)], line, column) if pointer.kind != "Pointer"
      return Type::NIL if pointer.inner.accepts?(value)

      type_error("type/set-deref-into-mismatch", [name_of(pointer.inner), name_of(value)], line, column)
    end

    # @return [Type]
    def check_builtin_free(args, line, column)
      pointer = check_expr(args[0])
      return Type::NIL if pointer.kind == "Pointer"

      type_error("type/free-expects-pointer", [name_of(pointer)], line, column)
    end

    # The number of heap slots the collector freed.
    #
    # @return [Type]
    def check_builtin_collect(_args, _line, _column)
      Type::INTEGER
    end
  end
end
